File: backend/app/schemas/ask.py
from typing import Literal

from pydantic import BaseModel, Field


class AskRequest(BaseModel):
    question: str = Field(min_length=1)


class AskSource(BaseModel):
    owner_type: str
    owner_id: str
    score: float
    title: str
    raw_item_id: str | None = None


class AskResponse(BaseModel):
    ask_run_id: str | None = None
    answer: str
    sources: list[AskSource]


class AskSaveRequest(BaseModel):
    save_as: Literal["task", "open_question", "decision"]
    title: str | None = None
    body: str | None = None
    rationale: str | None = None
    confidence: float = Field(default=0.5, ge=0, le=1)


class AskSaveResponse(BaseModel):
    raw_item_id: str
    memory_id: str
    entity_type: str
    entity_id: str

File: frontend/src/api/ask.ts
import { api } from './client';

export type AskResponse = {
  ask_run_id: string | null;
  answer: string;
  sources: { owner_type: string; owner_id: string; score: number; title: string; raw_item_id: string | null }[];
};

export type AskSavePayload = {
  save_as: 'task' | 'open_question' | 'decision';
  title?: string;
  body?: string;
  rationale?: string;
  confidence?: number;
};

export type AskSaveResponse = {
  raw_item_id: string;
  memory_id: string;
  entity_type: string;
  entity_id: string;
};

export function askQuestion(question: string) {
  return api<AskResponse>('/ask', {
    method: 'POST',
    body: JSON.stringify({ question })
  });
}

export function saveAskResult(askRunId: string, payload: AskSavePayload) {
  return api<AskSaveResponse>(`/ask/${askRunId}/save`, {
    method: 'POST',
    body: JSON.stringify(payload)
  });
}
